File: architecture-skills/object-oriented-programming-paradigm/examples/composition-over-inheritance/composition-over-inheritance.cpp
// Composition over inheritance in C++.
//
// Instead of a virtual Robot hierarchy with one subclass per movement x
// weapon combo, Robot owns std::unique_ptr<Interface> members and
// delegates to them. Small interfaces are still used — but for the
// parts, not for the robot's identity. Parts swap at runtime via a
// setter, which no subclass could do.

#include <iostream>
#include <memory>
#include <string>
#include <vector>

struct MoveBehavior {
    virtual ~MoveBehavior() = default;
    virtual std::string move() const = 0;
};

struct AttackBehavior {
    virtual ~AttackBehavior() = default;
    virtual std::string attack() const = 0;
};

// --- Concrete components: one capability each ---

struct Treads : MoveBehavior {
    std::string move() const override { return "rolls forward on treads"; }
};
struct Rotors : MoveBehavior {
    std::string move() const override { return "lifts off on quad rotors"; }
};
struct Turbines : MoveBehavior {
    std::string move() const override { return "dives and swims with turbines"; }
};

struct Laser : AttackBehavior {
    std::string attack() const override { return "fires a precision laser"; }
};
struct Drill : AttackBehavior {
    std::string attack() const override { return "spins up a mining drill"; }
};

// Host: owns its parts and delegates behavior to them.
class Robot {
public:
    Robot(std::string name, std::unique_ptr<MoveBehavior> mover,
          std::unique_ptr<AttackBehavior> weapon)
        : name_(std::move(name)), mover_(std::move(mover)), weapon_(std::move(weapon)) {}

    void setMover(std::unique_ptr<MoveBehavior> mover) { mover_ = std::move(mover); }

    void perform() const {
        std::cout << name_ << ": " << mover_->move() << " and "
                  << weapon_->attack() << "\n";
    }

private:
    std::string name_;
    std::unique_ptr<MoveBehavior> mover_;
    std::unique_ptr<AttackBehavior> weapon_;
};

int main() {
    // Any combination is just wiring — no FlyingDrillRobot subclass needed.
    std::vector<Robot> squad;
    squad.emplace_back("Scout", std::make_unique<Treads>(), std::make_unique<Laser>());
    squad.emplace_back("Miner", std::make_unique<Treads>(), std::make_unique<Drill>());
    squad.emplace_back("Wasp", std::make_unique<Rotors>(), std::make_unique<Drill>());
    squad.emplace_back("Manta", std::make_unique<Turbines>(), std::make_unique<Laser>());

    for (const Robot &robot : squad) {
        robot.perform();
    }

    // Inheritance fixes behavior at construction; composition swaps it live.
    std::cout << "-- field upgrade: Miner receives rotors --\n";
    squad[1].setMover(std::make_unique<Rotors>());
    squad[1].perform();
}

File: architecture-skills/object-oriented-programming-paradigm/examples/composition-over-inheritance/composition-over-inheritance.py
"""Composition over inheritance in Python.

A robot hierarchy would need one subclass per movement x weapon combo
(WalkingLaserRobot, FlyingDrillRobot, ...). Composing a Robot from
swappable parts needs 3 + 2 components for all 6 combos — and parts can
be replaced at runtime. Protocol gives static checking without a base
class; duck typing alone would also do.
"""

from __future__ import annotations

from typing import Protocol


class MoveBehavior(Protocol):
    def move(self) -> str: ...


class AttackBehavior(Protocol):
    def attack(self) -> str: ...


# --- Concrete components: one capability each, independently testable ---


class Treads:
    def move(self) -> str:
        return "rolls forward on treads"


class Rotors:
    def move(self) -> str:
        return "lifts off on quad rotors"


class Turbines:
    def move(self) -> str:
        return "dives and swims with turbines"


class Laser:
    def attack(self) -> str:
        return "fires a precision laser"


class Drill:
    def attack(self) -> str:
        return "spins up a mining drill"


class Robot:
    """Host object: owns its parts and delegates behavior to them."""

    def __init__(self, name: str, mover: MoveBehavior, weapon: AttackBehavior) -> None:
        self.name = name
        self.mover = mover
        self.weapon = weapon

    def perform(self) -> str:
        return f"{self.name:>8}: {self.mover.move()} and {self.weapon.attack()}"


if __name__ == "__main__":
    # Any combination is just wiring — no FlyingDrillRobot subclass needed.
    squad = [
        Robot("Scout", Treads(), Laser()),
        Robot("Miner", Treads(), Drill()),
        Robot("Wasp", Rotors(), Drill()),  # flying driller: new subclass in a hierarchy
        Robot("Manta", Turbines(), Laser()),
    ]
    for robot in squad:
        print(robot.perform())

    # Inheritance fixes behavior at construction; composition swaps it live.
    print("-- field upgrade: Miner receives rotors --")
    squad[1].mover = Rotors()
    print(squad[1].perform())
